org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/BranchLabel.java

```diff
@@ -116,7 +117,39 @@ public void becomeDelegateFor(BranchLabel otherLabel) {
 	this.forwardReferences = mergedForwardReferences;
 	this.forwardReferenceCount = indexInMerge;
 }
-
+private boolean shouldThrow = true;
+protected void trackStackDepth(boolean branch) {
+	/* Control can reach an instruction with a label in two ways: (1) via a branch using that label or (2) by falling through from the previous instruction.
+	   In both cases, we KNOW the stack depth at the instruction from which control flows to the instruction with the label.
+	   Control cannot reach the instruction with a label by falling through if the previous instruction completes abruptly via a goto/return/throw etc
+	   this.codeStream.lastAbruptCompletion == this.position ==> instruction prior to the one with the label is a goto/return/throw
+	*/
+	boolean sourceDepthKnown = branch || this.codeStream.lastAbruptCompletion != this.position;
+	if (this.targetStackDepth == -1) {
+		// First branch to this label || placement of this label
+		if (sourceDepthKnown) {
+			if (this.codeStream.stackDepth < 0) {
+				this.codeStream.classFile.referenceBinding.scope.problemReporter()
+						.operandStackSizeNegative(this.codeStream.classFile.referenceBinding.scope.referenceContext);
+				this.codeStream.stackDepth = 0; // fwiw
+				if (this.shouldThrow)
+					throw new AssertionError("Operand stack size negative!"); //$NON-NLS-1$
+			}
+			this.targetStackDepth = this.codeStream.stackDepth;
+		} // else: previous instruction completes abruptly via goto/return/throw: We can't use the stack depth there; Wait for a backward branch to be emitted.
+	} else {
+		// Stack depth known at label having encountered a previous branch and/or having fallen through to label
+		if (sourceDepthKnown) {
+			if (this.targetStackDepth != this.codeStream.stackDepth) {
+				this.codeStream.classFile.referenceBinding.scope.problemReporter().operandStackSizeInconsistent(
+						this.codeStream.classFile.referenceBinding.scope.referenceContext);
+				if (this.shouldThrow)
+					throw new AssertionError("Operand stack size inconsistent!"); //$NON-NLS-1$
+			}
+		}
+		this.codeStream.stackDepth = this.targetStackDepth;
```

Comment:
As a safety net, shouldn't this be max of the two values [assume sourceDepth is false and this.targetStackDepth != this.codeStream.stackDepth
[interim review comment]

Reply:
Agreed, I plan to remove the assertion - have only warning and can pick the higher of any two differing values as a safety option,